### PR TITLE
Fix the sampler truncation strategy

### LINKED ISSUE #712: Sampler should not blindly truncate out everything after the first end_token
The current behavior is truncating out everything after first seeing `end_token`, which is not always the right move. For example, in a chatbot generation system, each conversation segment will have a appended end token, so we only want to truncate the generated text. For example, if we have this prompt "what's up? <|endoftext|> Nothing special <|endoftext|> what something cool? ", we should not truncate everything after "what's up".

The proposed fix is to truncate everything after end token is seen within generated text only.

### PR DESCRIPTION
**Pull request by chenmoneygithub**
resolve #712

### CODE REVIEW

File: keras_nlp/samplers/sampler.py [hunk 1]
```diff
@@ -189,14 +189,27 @@ def _pad_prompt(self, prompt, max_length):
...
+        generated_result,
+        original_prompt,
         max_length,
         end_token_id,
     ):
         """Helper function to truncate the tokens after the end token."""
+        # Exclude original prompts from being truncated.
+        # Add a big int to tokens in original prompt to move token ids over
+        # the vocab size. Vocab size should not be large as 1e9.
+        increment_value = int(1e9)
```
Comment (mattdangerw) [inline keras_nlp/samplers/sampler.py:201]: these feels a little hacky. let's try to think of a more elegant way to do this. maybe pass in the original mask after padding and do something like this...

```python
# Create a tensor with True for each end_token_id
end_tokens = generated_result == end_token_id
# Remove all end_token_ids in the original input.
end_tokens = end_tokens & (original_padded_mask == False)
# Find index of first end_token_id.
end_indices = tf.math.argmax(generated_result == end_token_id, -1)
```

**Review by mattdangerw — COMMENTED** [review]
Left a few comments. Overall the thing that is still striking me about this design is how special cased the ragged code for doing the pre and post processing is getting (even in token id space). Seems like we have a few phases for generation...

1) tokenize inputs
2) mask and pack inputs
3) sample
4) truncate and manipulate outputs
5) detokenize

It seems unlikely that we will ever do 2) and 4) in a way that is perfectly general, so it's unfortunate that 2), 3), and 4) all come together. It does not feel very modular.

Not something we need to solve on this PR, but let's keep thinking about that.

**Comment by chenmoneygithub** [comment]
@mattdangerw Yea this sampler implementation now contains a lot of things, including preprocessing, sampling algo and postprocessing. It's a bit hard to figure out the best user flow with the class itself, we could keep thinking about it while we are rolling out the support for GPT2 text generation, including cache, long sentence generation and so on.

File: keras_nlp/samplers/sampler.py [hunk 2]
```diff
@@ -228,6 +231,7 @@ def __call__(
         # static shape, which means we cannot concatenate generated token to
         # current prompt.
         prompt, mask = self._pad_prompt(prompt, max_length)
+        original_padding_prompt = tf.identity(mask)
```
Comment (mattdangerw) [inline keras_nlp/samplers/sampler.py:234]: Can we use the same name here as above? `original_padded_mask`

**Review by mattdangerw — APPROVED** [review]
Thanks!

**Pull request merged**